### PR TITLE
first cut at 0.7 compatibility

### PR DESCRIPTION
**Pull request by sbromberger**
This requires the `m3/where-fns` branch of `SimpleTraits.jl`.

Current status:
```
Test Summary:              |  Pass  Error  Total
LightGraphs                | 28668      5  28673
  LightGraphs.SimpleGraphs |  2730          2730
  LightGraphs.LinAlg       |   336      4    340
    Graph matrices         |    45      1     46
    Spectral               |   291      3    294
  Interface                |    23            23
  Core                     |   122           122
  Operators                |   318           318
  Decomposition            |    48            48
  Distance                 |    42            42
  Transitivity             |    24            24
  Hadwick-James Cycles     |    21            21
  Cycles                   |    30            30
  Karp Minimum Cycle Mean  |    30            30
  Edit distance            |   240           240
  Connectivity             |   198      1    199
  Persistence              |    15            15
  Randgraphs               |  1270          1270
  Static graphs            |    25            25
  Small graphs             |    25            25
  Euclidean graphs         |    13            13
  A star                   |    18            18
  Bellman Ford             |    48            48
  Dijkstra                 |   441           441
  Floyd Warshall           |    21            21
  Yen                      |    72            72
  BFS                      |    75            75
  Parallel_BFS             |    25            25
  Bipartiteness            |    27            27
  DFS                      |    24            24
  Max adj visit            |    33            33
  Random walks             | 21879         21879
  Diffusion Simulation     |    57            57
  Cliques                  |    12            12
  Core periphery           |    93            93
  Label propagation        |    24            24
  Modularity               |     6             6
  Clustering               |    18            18
  Clique percolation       |     1             1
  Betweenness              |    51            51
  Closeness                |    48            48
  Degree                   |     9             9
  Katz                     |     3             3
  Pagerank                 |    54            54
  Eigenvector              |     6             6
  Stress                   |    18            18
  Radiality                |     6             6
  Utils                    |    18            18
  Kruskal                  |     6             6
  Prim                     |     6             6
  Articulation             |    24            24
  Biconnect                |     6             6
  Normalized Cut           |    29            29
```

### CODE REVIEW

**Comment by sbromberger** [comment]
I've asked Mauro to tag a new version of SimpleTraits that has the where-fn fix in it. Once we have that I'll make it part of REQUIRES and will retest.

File: test/runtests.jl [hunk 1]
```diff
@@ -6,6 +6,9 @@ using LinearAlgebra
 
 const testdir = dirname(@__FILE__)
 
+# we need a specific branch of SimpleTraits.jl until https://github.com/mauro3/SimpleTraits.jl/pull/51 is merged
+Pkg.checkout("m3/where-fns")
```
Comment (rohitvarkey) [inline test/runtests.jl]: I think you need `Pkg.checkout("SimpleTraits", "m3/where-fns")`

File: src/LightGraphs.jl [hunk 2]
```diff
@@ -1,15 +1,28 @@
 __precompile__(true)
 module LightGraphs
 
+using SharedArrays
+using Random
```
Comment (sbromberger) [inline src/LightGraphs.jl]: We will want to dedup these.

**Comment by rohitvarkey** [comment]
Not sure why Travis is failing. This is what I get on my local machine with `julia` built with the latest master and this branch.
 
```julia
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: https://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.7.0-DEV.4031 (2018-02-20 14:42 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit 5aa548b453* (0 days old master)
|__/                   |  x86_64-apple-darwin16.4.0

julia> Pkg.test("LightGraphs");
*....deprecations.....*
Test Summary: |  Pass  Total
LightGraphs   | 28845  28845
[ Info: LightGraphs tests passed
```

**Comment by sbromberger** [comment]
Looks like it's still not using the correct branch of SimpleTraits.

**Comment by sbromberger** [comment]
https://github.com/JuliaLang/METADATA.jl/pull/13453

**Comment by sbromberger** [comment]
OK, so: should we merge this now, or try to get rid of some of these deprecations first?

If we merge now, we'll need to cut a final 0.6-compatible version beforehand. This will allow us to use FemtoCleaner on the master branch.

**Comment by rohitvarkey** [comment]
I tried setting up FemtoCleaner locally to try and remove deprecations on this branch but had lots of issues with Homebrew etc. 

Considering that, I think the approach of cutting a final 0.6 release, merging this PR and then running FemtoCleaner on master is probably the best way of removing the deprecations and getting LG completely ready for 0.7.

**Comment by sbromberger** [comment]
https://github.com/JuliaGraphs/LightGraphs.jl/releases/tag/v0.12.0

https://github.com/JuliaLang/METADATA.jl/pull/13458

**Comment by rohitvarkey** [comment]
Hmm. Seems like Pkg isn't finding the correct version of SimpleTraits. Could this be cause SimpleTraits is set as `julia 0.7` and not `julia 0.7-` in METADATA?

**Comment by sbromberger** [comment]
> Could this be cause SimpleTraits is set as julia 0.7 and not julia 0.7- in METADATA?

I'll bet that's it.

**Comment by rohitvarkey** [comment]
Now that the METADATA PR is merged, let's try Travis again? 🤞

**Comment by sbromberger** [comment]
OK, tests pass. I'm going to merge this into master.

**Pull request merged**